compiler/rustc_hir_typeck/src/callee.rs

```diff
@@ -135,6 +138,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         output
     }
 
+    /// Functions of type `extern "custom" fn(/* ... */)` cannot be called using `ExprKind::Call`.
+    /// These functions have a calling convention that is unknown to rust, hence it cannot generate
+    /// code for the call. The only way to execute such a function is via inline assembly.
+    fn check_call_custom_abi(&self, callee_ty: Ty<'tcx>, span: Span) {
+        let abi = match callee_ty.kind() {
+            ty::FnDef(def_id, _) => self.tcx.fn_sig(def_id).skip_binder().skip_binder().abi,
+            ty::FnPtr(_, header) => header.abi,
+            _ => return,
```

Comment:
Are there other valid callees? If so, it may be good to `return` from them specifically then `bug!` on the fallback `_`.

Reply:
Hmm, that raises the question of what `async extern "custom"  fn foo() {}` should do. It always returns `impl Future`, so I think we should just disallow this.

Reply:
Agreed to disallowing async 👍 that's likely pretty tied to the Rust ABI too.

Reply:
I believe that what can happen here is that the scrutinee is some `impl Fn()` (or similar) value: it is callable, but not a function definition or function pointer. 

@oli-obk you've been working in this area a bunch: do you have advice on getting the ABI here (anything that implicitly uses the `extern "Rust"` (closures, `Fn()`) is irrelevant, we're just looking for callable items with the `"custom"` ABI. I suspect `&unsafe extern "custom" fn()` in particular is going to be tricky and is missed by the current logic. Is there some standard way to get at the "real" type?

Reply:
autoderef has already happened at this point. You can invoke `autoderef.final_ty()` to get the actual type that we'll be invoking things on. That should resolve any concerns about `&T`. Since Nothing but function pointers and function definitions is callable *and* can have a fn abi other than `"Rust"`, the current code otherwise seems fine. Some tests are good regardless. So basically just ensuring that you can't use the "custom" abi in impl methods or trait methods and calling references to function pointers or even function definitions.

Reply:
Ah, that's helpful! I'm still hitting `ty::Alias` and `ty::Adt` in this match though (maybe these become errors later). Can we safely ignore those cases with the `_ => return`?

Reply:
Yea there is no way to make them have any abi but Rust